tests/solana.rs

```diff
@@ -39,6 +36,20 @@ use tiny_keccak::{Hasher, Keccak};
 mod borsh_encoding;
 mod solana_tests;
 
+/// Error handling for syscall methods
+macro_rules! question_mark {
```

Comment:
I think a much more idiomatic way to do this would be to return an `Error` instead of passing a mutable pointer to some value representing an error. Then you can either use `map_err` with the actual question mark `?` syntax sugar

```rust
computation = whatever().map_err(|e| ProgramResult::Err(e.into()))?;
```

Or, even better, the casting can be directly implement, something like `impl From<WhatEverError> for ProgramResult`. Which would no longer require `map_err` and the `?` sugaring should just work.

Completely unrelated to what the PR is trying, but I couldn't resist sorry :monkey: So feel free to just take this as an improvement / suggestion that can be at a later time anyways.

Reply:
> I think a much more idiomatic way to do this would be to return an `Error` instead of passing a mutable pointer to some value representing an error.

Yes, I totally agree. Unfortunately the system call interface of the `solana_rbpf` crate takes `result` as a mutable pointer and returns unit, so there is nothing we can do about it,

I don't really know why that is done. If I were to hazard a guess, it is to avoid an allocation of a `Result<>` in the system call. I can't find an explanation in the source code.

> Completely unrelated to what the PR is trying to achieve, but I couldn't resist, sorry monkey So feel free to just take this as an improvement / suggestion that can be at a later time anyways.

It's a good point, I'll ask why it has been done this way.

Reply:
I see. Assuming they have their own fair reason, would be interesting to know :blush:

Reply:
I got an answer: That has to do with the calling convention and wether arguments and return values are passed as registers or through the stack. By using a reference we can explicitly control that, thus the ABI is more stable.